### PR TITLE
Add `SyncFinished` message to the network protocol

### LINKED ISSUE #3173: Extend the networking protocol with SyncStateChanged
The nodes that are participating may still be running the "sync to genesis" process in the background. Until the background syncing process is finished we shouldn't query such node for tries.
Proposed extension to the network protocol:
- rename the current `is_joiner` flag to `sync_state` flag (includes the related code, like `update_joining_set`, etc.)
- introduce an announcement for flipping the `sync_state` flag
  - extend small_network::Message by adding new `SyncStateChanged(bool)` variant
  - consider allowing the one side flipping (from true to false) as node should never switch from "participating" back to "joiner"
- decouple the handshake from setting the syncing state flag
  - remove the `is_joiner` from the `Message::Handshake` variant
  - on handshake default the flag to `true`
  - send the `SyncStateChanged` announcement with correct state afterwards
- the sync_state flag should be set to true only in these two conditions:
  - node is in the joiner reactor stage
  - node is in the participating reactor, but is doing the sync_to_genesis in the background
- ~upon receiving the SyncStateChanged(true), disconnect from all other nodes currently known as being in sync state (formerly, joiners)~ - there will be just a `SyncFinished` flag, meaning we'll never change the sync status back to true, hence no need for disconnection

- ~consider changes to the currently existing NodeState struct to better describe what mode the node is in at the given moment~ separate ticket will be requested because we need to discuss the actual states we now have in the node

### PR DESCRIPTION
**Pull request by rafal-ch**
This PR changes the following:

- Introduces the `ChainSynchronizerAnnouncement` with `SyncFinished` variant. It is announced in two cases

  -  when node is not configured to `sync_to_genesis` it is sent immediately after the participating reactor spins up the chain synchronizer component
  - otherwise, it is sent after the `sync_to_genesis` is completed

- Makes the `small_network` react to the `SyncFinished` event by disconnecting all incoming connections and updating the internal `syncing` flag to `false`, so on reconnection, the `false` value will be used as an `is_syncing` parameter of the handshake

Closes #3173

### CODE REVIEW

**Comment by rafal-ch** [comment]
Adding the `do not merge` label, because some housekeeping needs to be done after https://github.com/casper-network/casper-node/pull/3167 is merged.

File: node/src/components/chain_synchronizer.rs
Comment (marc-casperlabs) [inline node/src/components/chain_synchronizer.rs]: ~We should be more specific what kind of sync we finished, since we have multiple ones (to genesis or fast sync) and I would like to avoid confusion. Just changing the name is probably sufficient.~

Let's put into the doc comments that it indicates the nodes has finished the kind of sync it was doing (which is not specified) and will cease to request unsafe data. Also mention that we rely on reconnecting to reset the syncing state.

Comment (marc-casperlabs) [reply node/src/components/chain_synchronizer.rs]: Also, the "one announcement per component" is a bit of a (widespread) anti-mattern, at least in my mind. This could just be a 

```rust
struct SyncFinishedAnnouncement;
```

instead.

File: node/src/components/small_network.rs
Comment (marc-casperlabs) [inline node/src/components/small_network.rs]: How about `syncing_nodes` instead?

File: node/src/components/small_network.rs
Comment (marc-casperlabs) [inline node/src/components/small_network.rs]: `sync_in_progress` or `synchronization_in_progress` is, I believe, a better term here? Maybe I'm being overly pedantic though.

File: node/src/components/small_network.rs
Comment (marc-casperlabs) [inline node/src/components/small_network.rs]: Please remove this and move the other net metrics incrementation above into `fn broadcast_message` instead, this is way too error prone. I believe there should only be a single call to `.inc()` for almost any metric, along with one for `.dec()`.

File: node/src/components/small_network.rs
Comment (marc-casperlabs) [inline node/src/components/small_network.rs]: It does not make a difference right now, but I'd prefer the order be swapped (set the flag first, then broadcast) just in case.

File: node/src/reactor/participating.rs
Comment (marc-casperlabs) [inline node/src/reactor/participating.rs]: I think we should keep this in line with the rest of the code and implement `From<ChainSynchronizerAnnouncement>` (or even better `From<SyncFinishedAnnouncement>`, see comments above), instead of having the conversion here. Component events are supposed to be opaque to the reactor.

File: node/src/components/small_network/event.rs
Comment (marc-casperlabs) [inline node/src/components/small_network/event.rs]: I think turning this into a `ChainSynchronizerAnnouncement(_)` (see comment below) or `SyncFinishedAnnouncement(_)` variant is better - with the proposed changes we would have to implement `From<_>` manually here.

Also I have no idea what `#[from]` does with the code as-is :)

File: node/src/components/small_network/message.rs
Comment (marc-casperlabs) [inline node/src/components/small_network/message.rs]: We do not allow a second handshake, thus we can get away with a hardcoded weight of 0 here. However, we should have a weight for `SyncFinished`, or we need to reject/disconnect whenever a client sends a `SyncFinished` message when it already has synced, to close this DOS vector.

File: node/src/components/small_network/tasks.rs
Comment (marc-casperlabs) [inline node/src/components/small_network/tasks.rs]: I think we should remove this parameter entirely, as it duplicates state already found in the component. I made a mistake in adding it here a while ago, sorry!

Instead, we should put the check in the component itself, into the `send_message` function just before the call `connection.sender.send((msg, opt_responder))`. That way, information will always be up-to-date and no potentially stale state duplicated in the sender task.

It will be a tiny bit more complex (more hash set membership checks), but that is negligible overhead I think. If it turns out to be a problem, we can add it to the outgoing connection handle instead, but we expect this feature to go away anyway.

File: node/src/protocol.rs
Comment (marc-casperlabs) [inline node/src/protocol.rs:146]: We have a trait describing the same information now in the chain synchronizer, which we need to clean up (`CanUseJoiners`).

I think it'd be best to put the `CanUseJoiners` functionality onto the `Tag` for now:

```rust
impl Tag {
    fn can_use_joiners(self) -> bool;
}
```

For any function handling an item `T: Item`, we can then use `(<T as Item>::Tag).is_unsafe_for_joiners()`.

**Review by marc-casperlabs — DISMISSED** [review]
I'd like to see some changes, the bare minimum being the "is syncing" bool being removed from the sender task.

File: node/src/components/chain_synchronizer.rs
Comment (marc-casperlabs) [reply node/src/components/chain_synchronizer.rs]: ~~I believe there is technically now a race condition where the following happens:~~

~~* Node finishes syncing, sends "syncing finished message" to peers.~~
~~* Reactor shuts down (reactor switch). A ton of trie requests are queued at the right moment and stuck long enough in their event queues to move past the reconnection (again, this is because we do not refresh node IDs on reactor change anymore).~~
~~* New reactor again syncs because sync to genesis was requested and it immediately sends the trie requests to peers.~~
~~* -> potential deadlock, as we have two peers backpressuring due to large amounts of outstanding requests.~~

~~This could be alleviated by not sending the syncing finished announcement ever, if we know archival sync is coming next.~~

Nvm, it the check is here :) https://github.com/casper-network/casper-node/pull/3179/files#diff-bc770989714a898b4e4f4932bfddacc6787c9d1e9c1cc133e4ea44fbdac7d606R148

File: node/src/reactor/participating.rs
Comment (goral09) [reply node/src/reactor/participating.rs]: They're not opaque just b/c we implement `Into`.

File: node/src/components/small_network/tasks.rs
Comment (rafal-ch) [reply node/src/components/small_network/tasks.rs]: Done in 3568021397b436d142f9ec6a0c6c98db6e92a919

File: node/src/components/small_network.rs
Comment (rafal-ch) [reply node/src/components/small_network.rs]: Updated in b6b3b3af763e02c445c408fab3506b1dcb62c548.

File: node/src/components/small_network.rs
Comment (rafal-ch) [reply node/src/components/small_network.rs]: Updated in b6b3b3af763e02c445c408fab3506b1dcb62c548.

File: node/src/components/small_network.rs
Comment (rafal-ch) [reply node/src/components/small_network.rs]: Right! Updated in 79c78a21cbb81f79e48ef6d09ff118dab281041f.

File: node/src/components/small_network/message.rs
Comment (rafal-ch) [reply node/src/components/small_network/message.rs]: Ok, I'll disconnect from peer if it sends duplicated `SyncFinished` messages.

File: node/src/protocol.rs
Comment (rafal-ch) [reply node/src/protocol.rs:146]: I'd rather postpone this change to another PR to avoid scope creep as neither the trait nor the function was introduced as part of this PR.

File: node/src/components/chain_synchronizer.rs
Comment (rafal-ch) [reply node/src/components/chain_synchronizer.rs]: More explanatory comment added in 706e01007dbdf0eb44f9096de0e5159bda54446b, although I prefer to stick to the "one announcement per component" approach for now if you're not feeling strongly about it.

File: node/src/components/small_network.rs
Comment (rafal-ch) [reply node/src/components/small_network.rs]: As you prefer: eba559e347d5cef0e137c95972e795280bfafecc

File: node/src/components/small_network/event.rs
Comment (rafal-ch) [reply node/src/components/small_network/event.rs]: For now I removed the superfluous `#[from]` attribute and added the missing comment here: d10a025b33873ae906d6851c913294354b5651c9

I'll probably consider changing this to `ChainSynchronizerAnnouncement` though. Please don't resolve yet.

File: node/src/components/small_network/message.rs
Comment (rafal-ch) [reply node/src/components/small_network/message.rs]: It turned to be somewhat tricky to detect whether the notification from a given peer is first one or subsequent one. Therefore I decided to use the generic throttling mechanism.

Changes in this commit: b39d1fe990180664995dd4b6ec2cbed23c8c7ea3

File: node/src/components/small_network/event.rs
Comment (rafal-ch) [reply node/src/components/small_network/event.rs]: Turned into proper announcement in eedcb1856b8ac54dee18404660b6ed061d5e7960.

File: node/src/reactor/participating.rs
Comment (marc-casperlabs) [reply node/src/reactor/participating.rs]: Emphasis on "supposed". If I had to design the system again, any of these announcements would take the form of 

```rust
pub struct FooAnnouncement(FooAnnouncementInner);

enum FooAnnouncementInner {
    // ...
}
```

File: node/src/effect.rs
Comment (Fraser999) [inline node/src/effect.rs]: We have more than one notion of syncing in the node - I think we should be specific about this:
```suggestion
    pub(crate) async fn announce_finished_chain_syncing(self)
```

File: node/src/effect.rs
Comment (Fraser999) [inline node/src/effect.rs]: I'm not sure we need this log line.  We already have `info!("finished chain sync to genesis");` inside the chain sync code, and the other callsite is when the user has disabled sync-to-genesis, so seeing this log message might be a bit confusing.

**Review by Fraser999 — COMMENTED** [review]
Partial review - I'll complete if you don't get enough approvals before then.

In the meantime, is it possible to have the small_network not advertise itself as syncing when constructed inside the participating reactor where sync-to-genesis is disabled?

File: node/src/components/chain_synchronizer.rs
Comment (marc-casperlabs) [inline node/src/components/chain_synchronizer.rs:127]: ~~If this is called in the joiner with sync to genesis disabled, won't this announce the sync completed immediately, even though we still need to fast sync?~~ nvm, I misssed the specialization.

File: node/src/components/small_network.rs
Comment (marc-casperlabs) [inline node/src/components/small_network.rs]: This is not correct - we should _not_ be triggering the shutdown sender in any case anymore except when actually shutting down.

File: node/src/components/small_network.rs
Comment (marc-casperlabs) [inline node/src/components/small_network.rs]: Let's not commit these debug messages here, they will spam the logs. There is detailed information about connection states already available in the outgoing connection statemachine handling.

File: node/src/components/small_network.rs
Comment (marc-casperlabs) [inline node/src/components/small_network.rs]: Please lower these to debug level, otherwise SRE will hunt me down and make me fix it.

File: node/src/components/small_network.rs
Comment (marc-casperlabs) [inline node/src/components/small_network.rs]: This is not a message we should put out - it belongs in fast sync.

I'd support a (potentially debug) style message that states that the networking component was _informed_ about sync completion though.

File: node/src/components/small_network/message.rs
Comment (marc-casperlabs) [inline node/src/components/small_network/message.rs]: Maybe just "Whether the node is syncing." ?

File: node/src/effect.rs
Comment (marc-casperlabs) [reply node/src/effect.rs]: Regardless, I'd prefer we don't put log messages into the effect builder if we can avoid it.

**Review by marc-casperlabs — COMMENTED** [review]
Still some question :)

Overall we need to really watch what log messages we put into networking, as production will see these 2 orders of magnitude more often then devs in testing (we get complaints about spammy logs). Anything `INFO` and above should be used very sparingly!

**Comment by marc-casperlabs** [comment]
> In the meantime, is it possible to have the small_network not advertise itself as syncing when constructed inside the participating reactor where sync-to-genesis is disabled?

It could, it'd just add more logic. The current approach just reused the functionality we need anyway, and the savings is a 1-second reconnection cycle from clients, so we opted for that. As a nice side effect, we exercise the otherwise rarely hit code path :)

File: node/src/protocol.rs
Comment (marc-casperlabs) [reply node/src/protocol.rs:146]: Fair enough!

File: node/src/components/small_network/message.rs
Comment (marc-casperlabs) [reply node/src/components/small_network/message.rs]: Works for me for now!

**Review by Fraser999 — APPROVED** [review]
As discussed elsewhere - optimistic approval by @marc-casperlabs and myself given the outstanding comments are addressed.

File: node/src/effect.rs
Comment (rafal-ch) [reply node/src/effect.rs]: Updated in 5b91f0f74616a4c60822d297c58f08523917ec38

File: node/src/effect.rs
Comment (rafal-ch) [reply node/src/effect.rs]: Log message removed in 5b91f0f74616a4c60822d297c58f08523917ec38

File: node/src/components/small_network.rs
Comment (rafal-ch) [reply node/src/components/small_network.rs]: Change to the comment reverted in 5b91f0f74616a4c60822d297c58f08523917ec38

File: node/src/components/small_network.rs
Comment (rafal-ch) [reply node/src/components/small_network.rs]: Comment removed in 5b91f0f74616a4c60822d297c58f08523917ec38

File: node/src/components/small_network.rs
Comment (rafal-ch) [reply node/src/components/small_network.rs]: Done in 5b91f0f74616a4c60822d297c58f08523917ec38

File: node/src/components/small_network.rs
Comment (rafal-ch) [reply node/src/components/small_network.rs]: I removed the message in 5b91f0f74616a4c60822d297c58f08523917ec38 as it indeed doesn't bring much value.

File: node/src/components/small_network/message.rs
Comment (rafal-ch) [reply node/src/components/small_network/message.rs]: Comment changed in 5b91f0f74616a4c60822d297c58f08523917ec38

**Pull request merged**
